Add more uses for mousewheel scrolling

tothxa opened the pull request:
- map scrolling with touchpad
- game speed
- editor toolsize
- slider and spinbox UI elements
- soldier capacities and attacker count
- input queue

Also change parameters passed to mousewheel handlers.
Includes config file options and new Scrolling tab in Options menu.

tothxa committed:
Add configuration file options for mousewheel uses

tothxa committed:
Change arguments passed for mousewheel handling

tothxa committed:
Add new uses for mousewheel:

- slider
- spinbox
- input queue
- soldier capacities
- number of soldiers for attack
- game speed
- editor toolsize
- make minimap pass wheel events to main window

tothxa committed:
Add mousewheel scrolling options tab to options menu

remove `ctrl_zoom` option

tothxa commented:
See https://www.widelands.org/forum/topic/5349/ and https://www.widelands.org/forum/topic/5384/

I expect there will be a lot of change requests, as most of this is way beyond my programming knowledge...

tothxa committed:
codecheck fix

tothxa committed:
'src/ui_fsmenu/mousewheel_options.cc' was automatically formatted.

Noordfrees reviewed (commented):
Wow, great! A few first comments inline.

---

The pointing device selector in the options menu seems unintuitive though. It is not clear whether this refers to the only device that may be used for scrolling, or whether we can have separate settings for each device. Perhaps clarify in the tooltip.

Assuming it's the latter, then switching the pointing device discards all changes I've made so far. I select the modifiers to use for the external mouse, then choose the touchpad, set its settings, switch back to external mouse to check – and all changes are reset to the initial value. Argh. They should be kept in memory until explicitly reset.

The Reset All button could also do with a tooltip to explain what it does – does it reset all modifiers or all devices?

tothxa commented:
> The pointing device selector in the options menu seems unintuitive though. It is not clear whether this refers to the only device that may be used for scrolling, or whether we can have separate settings for each device. Perhaps clarify in the tooltip.

Yeah, it may not be the best idea after all with the current set of options. While ctrl+vertical for zoom / ctrl+horizontal for speed/toolsize may be desirable for efficiency, it is clearly not fit to be the default. I also dropped the unified inputqueue idea (you can still see it in the original mousewheel branch), where I considered setting horizontal to quantity and vertical to priority, but it turned out in testing to be just as confusing as it was useful. So now most defaults are the same, only zoom and map scroll change with it.

Maybe I should drop the whole dual default thing (could be handled in the dialog), and I'm not sure about the pointing device dropdown either. It's just that I wanted to make it clear when it makes sense to turn on map scrolling... but this seems to be too confusing in other ways.
May I change ctrl_zoom to be the default, so that turning on map scrolling becomes a single step even without the pointing device dropdown?

It is also a problem that SDL doesn't make it possible to tell whether the hardware is capable of horizontal scrolling: it doesn't tell (at least in my config) whether there are more than one devices, and which of them is sending the current event, and it doesn't even tell (in my config) when the system scroll direction is flipped.

> Assuming it's the latter, then switching the pointing device discards all changes I've made so far. I select the modifiers to use for the external mouse, then choose the touchpad, set its settings, switch back to external mouse to check – and all changes are reset to the initial value. Argh. They should be kept in memory until explicitly reset.

I wasn't sure how to handle this. If you think it makes sense to keep very different settings for single wheel and touchpad, it could be easily done in 2 separate config sections (say [mousewheel] and [touchpad_scrolling]), and then indeed, the dropdown would get a clear purpose.

I considered keeping track of which values were changed but not applied, but skipped it for now. I thought it would be overkill in most cases, when people would first select the type of pointing device they intend to use. If the above 2 sections idea were implemented, then it would definitely make sense to keep track of user changes for both kinds of devices. But it's probably best to scrap the whole pointing device type dropdown instead as well as the dual defaults.

BTW there's another issue coming from the general handling of config options: once the settings are applied, they all get written to the config file, so changing the pointing device doesn't change anything directly anymore, only the settings the reset button will choose (which will then get written in the file...) I think it should be possible to clear settings from the config file, but the code handling this is too complicated and spread out in too many places to me. This also affects keyboard shortcuts: resetting them pollutes the config file with a list of all defaults, and so later changes to defaults will not be followed anymore.

> The Reset All button could also do with a tooltip to explain what it does – does it reset all modifiers or all devices?

OK. (it resets modifiers and scroll directions including inverts to the settings appropriate for the selected device type) If I drop the device dropdown, this will also become more clear.

Noordfrees commented:
Ok then I think it would be best to drop the Device dropdown and have only one profile. Note that you can emulate all input device events through software (at least on X11) and even connect different types of input to each other, so attempting to decide what functionality is available on a given system is not always meaningful.

-1 for making Ctrl+Wheel the zooming default, this makes the feature harder to discover for new players. Perhaps you could enable map scrolling by default with the Ctrl+Shift modifier so it's easy to remap the modifiers without them getting in each other's way initially.

Lastly, I'd suggest to make the scrolling options pane a tab in the keyboard shortcuts window and rename that menu something like "Shortcuts & Modifiers" to keep related settings grouped.

tothxa committed:
First round of requested changes:
- remove 2D defaults
- change scrollbar to use common handler function
- convert #defines to constexpr
- don't static_cast int to bool

tothxa committed:
2 files were automatically formatted.

tothxa committed:
fight clang-format

tothxa committed:
remove superfluous include

tothxa committed:
Fix portability problem and compiler warnings
also some formatting adjustment

tothxa committed:
'src/wlapplication_mousewheel_options.cc' was automatically formatted.

tothxa committed:
fight clang-format

tothxa committed:
Merge branch 'master' into mousewheel-pr

tothxa committed:
Move scrolling options to keyboard shortcut window
also make keyboard shortcut window update size when main window is resized

tothxa committed:
fix layouting of apply_button_ in Scroll settings

tothxa committed:
Merge branch 'master' into mousewheel-pr

tothxa committed:
2 files were automatically formatted.

tothxa commented:
> Lastly, I'd suggest to make the scrolling options pane a tab in the keyboard shortcuts window and rename that menu something like "Shortcuts & Modifiers" to keep related settings grouped.

Done.

I did an unrelated change: I made the keyboard_options window resize with the main window (I needed it to test the layouting). Please let me know if it should be reverted from this PR.

Noordfrees commented:
> I did an unrelated change: I made the keyboard_options window resize with the main window (I needed it to test the layouting). Please let me know if it should be reverted from this PR.

Looks fine to me codewise, not tested yet.

---

Two more nits regarding code style:
- Please go easier on abbreviations ;) They tend to impact readability. E.g. `kApplyBtnW` should better be `kApplyButtonWidth` and `kMwDdMaxW` should be… should be……… (point made :wink:?)
Local variables that are used only for an intermediate result or temporary iterator can be called `x` or `t` or whatever, but anything with a larger scope should have a name that tells the reader immediately what the object does.
- When declaring `constexpr` or `static const` variables please use one declaration per statement, e.g.
```diff
- static const Sign2D  //
-    kDefaultSignValue(kSignIncreaseRight, kSignIncreaseUp),
-    kDefaultSignMove(kSignNextRight, kSignNextDown),  //
-    kDefaultSignScroll(kSignScroll, kSignScroll);
+ static const Sign2D kDefaultSignValue(kSignIncreaseRight, kSignIncreaseUp);
+ static const Sign2D kDefaultSignMove(kSignNextRight, kSignNextDown);
+ static const Sign2D kDefaultSignScroll(kSignScroll, kSignScroll);
```

tothxa committed:
fix constexpr declarations and names

tothxa committed:
Add translator hints

tothxa commented:
>     * Please go easier on abbreviations ;) They tend to impact readability. E.g. `kApplyBtnW` should better be `kApplyButtonWidth` and `kMwDdMaxW` should be… should be……… (point made wink?)

Sorry. It started out with all of them like the BtnW one, but went through some deletions/readditions/renamings, and the scope is really just one .cc.

>     * When declaring `constexpr` or `static const` variables please use one declaration per statement

Sorry again. I was afraid it would be so, and I can understand why. I just thought it's more readable without repeating the `constexpr int`-s.

tothxa committed:
Add handling of RTL languages

Noordfrees reviewed (commented):
Looking good now. I'm through with the review, only a few minor comments inline.

One more suggestion, it would be good to have a Reset To Defaults button for the mouse wheel options along with the Apply button.

tothxa committed:
replace preprocessor macros with inline functions

tothxa committed:
Add reset button for scrolling options

Noordfrees commented:
LGTM but there's a layouting glitch: Initially the reset/apply buttons are centered, but when I try to assign a modifier that conflicts with another modifier they suddenly jump to the left:

![grafik](https://user-images.githubusercontent.com/48293446/136022496-53b40b81-369c-4331-bba6-9cc18bad6d51.png)
![grafik](https://user-images.githubusercontent.com/48293446/136022448-ce26ed7d-4c54-49b7-9ca9-80d8d7026ffd.png)

tothxa commented:
I'd add a button to set recommended map scroll and zoom settings for touchpad between 'Scroll Map' and 'Change Game Speed', but I can't think of a nice short label (possibly explained by a tooltip) for it.

tothxa committed:
Fix layouting bug

tothxa committed:
remove superfluous include

tothxa committed:
Add button for recommended touchpad settings

tothxa committed:
'src/ui_fsmenu/mousewheel_options.cc' was automatically formatted.

tothxa committed:
Merge branch 'master' into mousewheel-pr

Noordfrees commented:
Layout works now. For the button position I'd suggest having them all on one line since the Touchpad button is essentially another Reset button:

<table>
<tr>
	<td><b>Label:</b>
	<td>Reset (Mouse)
	<td>Reset (Touchpad)
	<td>Apply
<tr>
	<td><b>Tooltip:</b>
	<td>Reset to the default settings recommended for standard mouses
	<td>Reset to the settings recommended for touchpads and horizontally scrolling mouses
	<td>Apply scroll settings
</table>

tothxa commented:
> Layout works now. For the button position I'd suggest having them all on one line since the Touchpad button is essentially another Reset button:
> Label: 	Reset (Mouse) 	Reset (Touchpad) 	Apply
> Tooltip: 	Reset to the default settings recommended for standard mouses 	Reset to the settings recommended for touchpads and horizontally scrolling mouses 	Apply scroll settings

As it is now, the touchpad button only sets zoom/scroll map, and asks for approval to set speed and toolsize in case of conflict. It never touches invert settings, while reset forces them all to default. Also, currently the touchpad button does not apply the settings. So it is a much gentler thing than the reset button.

And there is the big Reset All button in the bottom of the window common with the keyboard shortcuts, so I think it's important to have the more specific labels for reset and apply. Though fitting 3 wide buttons is only a matter of adding another box, or playing tricks with the *DirBox-es.

tothxa committed:
Move 'Set Touchpad' button to bottom button row

tothxa committed:
Merge branch 'master' into mousewheel-pr

Noordfrees commented:
Looks good, just one more minor alignment issue: The rows of dropdowns have vertical gaps of differing sizes (4px, 9px, and 16px) between them:
![grafik](https://user-images.githubusercontent.com/48293446/136219757-da941987-02bc-4c4c-98fb-67c43fa04783.png)
The 16px gap is probably intentional, but the 4px and 9px gaps should all be the same height.

tothxa commented:
> Looks good, just one more minor alignment issue: The rows of dropdowns have vertical gaps of differing sizes (4px, 9px, and 16px) between them: The 16px gap is probably intentional, but the 4px and 9px gaps should all be the same height.

I added the small dividers (9px = 1px + 2 * kPadding) with the touchpad button, to signify that the first 2 are set together while the other 2 are completely independent of each other. But yes, they are now probably more confusing than helpful.

tothxa committed:
Remove small dividers

Noordfrees reviewed (approved):
Great :) 
Let's have it when the builds are green
